### ISSUE TITLE
Feature Selection: Use criterion function instead of specific learner+resampling

### ISSUE THREAD

**Issue by filipre** [mlr-org/mlr#666]
Hey mlr community!

Currently, I'm working on implementing the filter-based feature selection algorithm "[LoKDFS](http://jmlr.csail.mit.edu/proceedings/papers/v25/azmandian12/azmandian12.pdf)" (Local Kernel Density Ratio-Based Feature Selection for Outlier Detection by Fatemeh Aymandian) in mlr. The algorithm tries different subsets of features and calculates for each subset a specific value by a given criterion function (see Page 4 [in the paper](http://jmlr.csail.mit.edu/proceedings/papers/v25/azmandian12/azmandian12.pdf) for the definition). The paper uses "sequential forward selection (SFS)" but one could think of other search techniques to create the set of features. The subset with the highest value will be selected as the feature subset.

From the documentation I read that it is possible to create custom filter-based feature selection. However, the only way to do that is to rank each feature and then to use the k-best features (alongside percentage values and thresholds). This is not exactly what I want because I have to rank a _subset_ and not a single feature.

Additionally, I read that it is also possible to create a wrapper-based feature selector. While you have to specify a search strategy (which is what I need), you have to specify a learner as well as a resample technique as well. As far as I know, the criterion function is not really a "learner" and does not rely on resampling as well. It just calculates a value based on the given dataset and features. I thought about implementing a custom learner but that sounds like an ugly hack (What resampling technique would I use? What would it "train"? Which category [classification, regression, ...] would it fall into? ...)

I could use [dropFeatures()](http://rpackages.ianhowson.com/cran/mlr/man/dropFeatures.html) but then I would have to implement the search technique on my own as well and that doesn't sound right either (since the work is already done, isn't it?). 

Does someone have an idea what the best way could be to implement such an algorithm in mlr? 

**tl;dr:** I would like to use the wrapper-based search strategies for a filter-based feature selection.

Thank you already!

**Comment by berndbischl** [comment]
Hi,
the problem is that what you are currently doing is not completely supported by the structure we have. 
What we currently can do is
a) support arbitrary filters where the filter assigns one value per feature
b) forward selection (and other search strategies) that evaluate the ML method on the selected set by resampling (eg holdout or cross-validation).

The approach b) is based on the stuff by Ron Kohavi. 
http://ai.stanford.edu/~ronnyk/wrappersPrint.pdf

In order to figure out how to help you, we need to know what you would like to give and take from an input/output type of perspective. 

So I assume, you would like to do approach b), but use you custome score function, from the LoKDFS paper? And reuse mlr's search strategies to optimize the feature set? Is that correct as a summary?

So you have a function 
calcFeatureSetScore(task, features) --> R

which is passed to the search method?

**Comment by filipre** [comment]
Hey, 

>So I assume, you would like to do approach b), but use you custome score function, from the LoKDFS paper? And reuse mlr's search strategies to optimize the feature set? Is that correct as a summary?

yes! That's a good summary. 

I could think of following input/output:

```R
# Input
makeFilterScore(
  name = "lokdfs",
  ...
  fun = function(task, features, ...) { 
    # calculate value based on task data and the given features. this can be anything.
  }
)

# ... define task ...
ctree <- makeLearner("classif.ctree", id = "ctree")
ctree.lokdfs <- makeFilterWrapper(learner = ctree, score.func = "lokdfs", score.control = makeFeatSelControlGA(), k = 5, width = 2)
# ... create a model, do some benchmarking, whatever...

# Output
# model of learner ctree.lokdfs. will have the features that maximize the LoKDFS score function
```

To be honest, for me it is quite hard to define consistent input/output since I do not know mlr good enough but I could think of following requirements:
- usage of different search strategies
- like the other filters/preproc methods: be able to wrap other learners (see example above).
- custom parameters (like "kernel width", etc -> depending on the algorithm)

> the problem is that what you are currently doing is not completely supported by the structure we have. 

Currently, I think I could be able to write some trivial learner without learnable variables that calculates _one_ score for the whole dataset (in the `trainLearner`) and then sets every sample's target to that specific score (in `predictLearner`) like this:

```R
# TODO: properties, package, ....
makeRLearner.regr.lofcf = function() {
  makeRLearnerRegr(
    cl = "regr.lofcf",
    ...
  )
}

trainLearner.regr.lofcf = function(.learner, .task, .subset, .weights = NULL,  ...) {
  # get all the data and calculate the score (e.g. 5.13483)
  return(...model with value 5.13483...) # TODO
}

predictLearner.regr.lofcf = function(.learner, .model, .newdata, ...) {
  # return the score that we have calculated above.
  return(5.13483) # TODO
}
```

and then do some weird resampling (maybe a testset with all zeros in the target) in order to use [selectFeatures](http://rpackages.ianhowson.com/cran/mlr/man/selectFeatures.html) but obviously, this is not what a learner should do.

**Comment by berndbischl** [comment]
Ok. Lets summarize again. 

1) We create a Filter object, maybe this needs a specific class to separate it from the univariate filters.

```
makeFilterScore(
  name = "lokdfs",
  ...
  fun = function(task, features, ...) { 
    # calculate value based on task data and the given features. this can be anything.
  }
)
```

2) We then want to run search algos on top of this. This kind of implies that we would like to run "selectFeatures". The first arg of selectFeatures is a learner, which would normally be resampled like in the Kovavi-wrapper approach. Instead we would pass such a FilterScore object as above. Basically everything then works as before, the only difference is that the search algo uses a different objective function. 

3) We then need to be able to contstruct a feature selection wrapper. This should also still work.

@filipre 
Can you please also tell me your motivation behind this thread / implementation?  And long term goal?
(If you dont want to write this here, feel free to email)

**Comment by filipre** [comment]
Looks good!

Regarding 1)
I think we don't even need the `features` parameter because that information could be already in the `task` parameter (like it is with custom learners in the [trainLearner](https://mlr-org.github.io/mlr-tutorial/release/html/create_learner/index.html) function)

To our background/implementation:
My team and I were evaluating mlr as a new framework in respect of feature selection and extensibility. So we tried to implement some new fs algorithms with mlr to become familiar with these algorithms and with mlr as well. 
Currently, there is no long term goal. Therefore, having this feature isn't that important for us either (would just have been cool if it was already possible). I'm not even sure if there are a lot of other a algorithms that work similar like LoKDFS.
